deploy/test-environments/fleet_api/src/upgrade_agents.py

```
@@ -0,0 +1,181 @@
+
+    Returns:
+        bool: True if the package policy reaches the desired version within the timeout,
+              False otherwise.
+    """
+    start_time = time.time()
+
+    while time.time() - start_time < timeout_secs:
+        policy_info = get_package_policy_by_id(cfg, policy_id)
+        policy_name = policy_info.get("name", "")
+        policy_version = policy_info.get("package", {}).get("version", "")
+        logger.info(
+            f"Integration: {policy_name}, current version: {policy_version}, desired version: {desired_version}",
+        )
+        if policy_version == desired_version:
+            return True  # Desired version reached
+
+        time.sleep(poll_interval_secs)  # Wait and poll again
+
+    return False  # Desired version not reached within the timeout
+
+
+def main():
+    """
+    Main linux agents upgrade flow
+    """
+    # Ensure that all packages are on the latest version
+    wait_for_packages_upgrade()
+
+    download_source_id = create_custom_agent_download_source()
+
+    if not download_source_id:
+        logger.error("Failed to create the agent download source.")
+        sys.exit(1)
+
+    linux_policies_list = update_linux_policies(download_source_id)
+    time.sleep(180)  # To ensure that policies updated
+    agents = get_agents(cfg=cnfg.elk_config)
+    linux_agent_ids = [agent.id for agent in agents if agent.policy_id in linux_policies_list]
+    for agent_id in linux_agent_ids:
+        action_id = bulk_upgrade_agents(
+            cfg=cnfg.elk_config,
+            agent_ids=agent_id,
+            version=cnfg.elk_config.stack_version,
+            source_uri=get_artifact_server(version=cnfg.elk_config.stack_version),
+        )
+
+        if not wait_for_action_status(
+            cfg=cnfg.elk_config,
+            target_action_id=action_id,
+            target_type="UPGRADE",
+            target_status="COMPLETE",
+        ):
+            sys.exit(1)
```

Log before sys exit. Would be better to raise an exception in any case, no?